### DISCUSSION TITLE
Still running v0.32.0 and no clue how to upgrade

### DISCUSSION THREAD

**Discussion by hmeerlo** [fluxcd/flux2#4611] [General] [answered]
I guess the title is pretty clear. All I can find info on is migrating from v1 to v2. But no info on how to migrate from pre v1. Can anybody give me some pointers?

**Comment by hmeerlo** [comment]
Ok, am I right that there never was a `v1`? But it was always called `v0.x,y` and that you refer to that as being v1?

**Comment by hmeerlo** [comment]
Or is the plot thickening and was there a completely seperate v1 thing and is my `v0.32.0` just a prelease of `v2` that actually skipped past `v1.x.y` and went straight to `v2`? Very very confusing.

**Comment by stefanprodan** [comment, marked as answer]
Please see: 
- https://github.com/fluxcd/flux2/releases/tag/v2.0.0
- https://fluxcd.io/flux/installation/upgrade/

**Reply by hmeerlo** [reply]
Yes thanks. it took me 4 hours to figure out this very confusing versioning. I now understand how it works.

**Reply by l7u7c** [reply]
Hi @hmeerlo,
Did you successfully upgrade FluxCD v0.32.0 to v2.x.x?
I am in the same situation as you, and I want to upgrade FluxCD v0.32.0 to v2.x.x. My resources include ImageUpdater, HelmRelease, HelmRepository, and ImagePolicy.
However, I'm confused about how to upgrade it.

**Reply by hmeerlo** [reply]
No, I am still stuck at an old version. There is just too little attention to backwards compatibility and upgrade procedures. It is a list of bullets to perform and then hope that your production system still works, fingers crossed. So I will only do this once it is an absolute must.

**Reply by siegenthalerroger** [reply]
@hmeerlo we've been using Flux since the v1 days and went through each v0.xx pre-release and are now at v2.3.0. It's actually quite simple if you stay on top of it, but essentially you're just going to have to go through each (pre)release and read the release notes and apply the necessary CR changes after each update. A direct jump (probably) isn't going to work, as you're probably using some CRD versions that were deprecated and subsequently removed before the current version.

**Reply by hmeerlo** [reply]
Hi @siegenthalerroger , thanks. It is comforting to know that other people have successfully gone through this journey!

**Reply by kingdonb** [reply]
Edit: it's really not such a big deal as all this text, sorry I wrote so much, but this post is badly informed and I want to help!

> https://github.com/fluxcd/flux2/releases#:~:text=status.webhookPath)-,Upgrade%20procedure,-Upgrade%20Flux%20from

Look in the Release notes for "Upgrade procedure" - this tells you everything you need to know each minor release. Currently this all fits on one GitHub Releases page. `v0.x` to `v2.0.0`, and `v2.x` to `v2.3.0` is really all, there are also some notes `v2.x` to `v2.2.0` - these are highlighted consistently, you should be able to upgrade easily from any release or pre-release of fluxcd/flux2, in order to get to current.

>  The best way to keep your Flux up to date is to use the Flux update action. You need to read the update notes for each minor version to be aware of feature changes, but honestly loads of attention is paid to backwards compatibility, and deprecations are signaled for at least 6 months, sometimes in the history of Flux even longer - we are really serious about GA and v1 for a Kubernetes API means no backwards incompatible changes for us.

https://fluxcd.io/flux/installation/upgrade/#upgrade-with-git

Note the tiny aside here "automated upgrades"

If you wish to receive a PR every time a Flux upgrade is available, that you can just merge to upgrade, then you will only have one release note to read at a time. It is no surprise that people struggle when they don't realize Flux's goal is automation, and fully embrace the docs opportunities to automate. This makes what you are struggling with so much easier! ❤️‍🩹😁

The vast majority of release notes do not contain breaking changes. Flux releases minor versions approximately 3 times a year, those are feature updates and they can move APIs forward and enact deprecations that have been announced in prior release notes. If you are upgrading one minor version then you only need to read the minor release notes, as those are the ones that can contain breaking changes. After 2.0.0, things get much easier since Flux v2 is GA and as of now all the APIs are Beta or greater it doesn't have backwards incompatible changes. Just upgrade to v2.0.0 and read the release notes, plus each of the three minor releases after that, to get the full picture. Most of the "breaking changes" notes in these are repetitive and redundant of each other because these upgrades are signaled long in advance, it's no surprise when deprecations are enacted, and it happens honestly quite infrequently. (Deprecations from Kustomize are frankly the worst, but we have to accept them if we are to stay current with upstream Kustomize.)

Patch releases, don't contain any breaking changes so you can always approve patch releases, they only fix bugs. This is according to semver as described on semver.org.

If you are upgrading flux2 from a very old prerelease version, you can just read 2.0.0 release notes, and the release notes of the minor version that you are upgrading to (hopefully that is the current version). Those two will contain all of the details you need to know about breaking changes. Flux has made very few breaking changes after the v1->v2 switch, all of prereleases on Flux v2 can be quickly upgraded to the latest version with just a few changes.

Finally...If you aren't on GitHub, or don't use GHA you can still get the same automated upgrade experience by enabling Renovate Bot. This one actually captures all the changelogs between the version that you are on, and the version to upgrade to, and posts them for you in the PR so you don't have to go back to release notes on GitHub. That works for Github and GitLab at least, probably others too.

We have an active community of volunteers on the CNCF slack where someone will help in #flux if you are stuck. It's a better forum for issues like this where you just get the error and need to get unstuck, we can help you move on and enjoy the latest GA Flux. Sorry for the inconveniences, just understand that according to semver prereleases (0.x.y) can have breaking changes on Minor versions, and GA releases (1.x.y, 2.x.y) cannot have breaking changes without a MAJOR bump. Flux2 spent a long time in prereleases to get everything right (that's why you see so many release notes!), so, well, we think Flux is perfect now, and so it isn't planning any new major bumps afaik 😉

So rest assured, once you upgrade to the latest version and set up automation to stay current, you won't struggle with backwards incompatible changes. By and large, you can just upgrade, and on the rare occasion something breaks, go back and read the release notes then fix it. Automation is your friend. Even without automated upgrades, the Flux bootstrap process works on any version to upgrade and is built to be an idempotent operation, so if it fails somehow, you can just fix it and try again.

Flux errors also don't make production outages unless you consider dev work as production. It's just stop and fix, then move on 🤓 it's not something that needs a lot of hand wringing, truthfully. I hope you were able to upgrade.
